File: frontend/lib/models/teacher.dart
import 'package:freezed_annotation/freezed_annotation.dart';

part 'teacher.freezed.dart';
part 'teacher.g.dart';

@freezed
class TeacherAssignment with _$TeacherAssignment {
  const factory TeacherAssignment({
    required String classId,
    required String subjectName,
  }) = _TeacherAssignment;

  factory TeacherAssignment.fromJson(Map<String, dynamic> json) => _$TeacherAssignmentFromJson(json);
}

@freezed
class Teacher with _$Teacher {
  const factory Teacher({
    required String id,
    required String madrasaId,
    required String staffId,
    @Default('') String firstName,
    @Default('') String lastName,
    @Default('') String phone,
    String? email,
    @Default([]) List<String> qualifications,
    @Default([]) List<TeacherAssignment> assignedClasses,
    @Default(false) bool isDeleted,
    @Default('') String fullName,
  }) = _Teacher;

  factory Teacher.fromJson(Map<String, dynamic> json) => _$TeacherFromJson(json);
}

File: backend/src/models/Teacher.js
const mongoose = require('mongoose');

const TeacherAssignmentSchema = new mongoose.Schema({
  classId: {
    type: mongoose.Schema.Types.ObjectId,
    ref: 'Class',
    required: true,
  },
  subjectName: {
    type: String,
    required: true,
    trim: true,
  },
});

const TeacherSchema = new mongoose.Schema(
  {
    madrasaId: {
      type: mongoose.Schema.Types.ObjectId,
      ref: 'Madrasa',
      required: true,
    },
    staffId: {
      type: String,
      required: true,
      trim: true,
    },
    firstName: {
      type: String,
      required: true,
      trim: true,
    },
    lastName: {
      type: String,
      required: true,
      trim: true,
    },
    phone: {
      type: String,
      required: true,
      trim: true,
    },
    email: {
      type: String,
      trim: true,
      lowercase: true,
    },
    qualifications: {
      type: [String],
      default: [],
    },
    assignedClasses: {
      type: [TeacherAssignmentSchema],
      default: [],
    },
    isDeleted: {
      type: Boolean,
      default: false,
      index: true,
    },
  },
  {
    timestamps: true,
  }
);

// Compound Unique Index: Enforce unique staff/teacher ID per madrasa
TeacherSchema.index({ madrasaId: 1, staffId: 1 }, { unique: true });

// Compound Index: Optimize teacher filtering by tenant context
TeacherSchema.index({ madrasaId: 1, isDeleted: 1 });

// Text Index: Optimize global keyword search for teacher names
TeacherSchema.index({ firstName: 'text', lastName: 'text' });
TeacherSchema.index({ madrasaId: 1, 'assignedClasses.classId': 1 });
TeacherSchema.index({ 'assignedClasses.subjectName': 1 });

module.exports = mongoose.model('Teacher', TeacherSchema);
